misc/graphify.js

```diff
@@ -1,122 +1,48 @@
-var crawler = require('../src/crawler.js');
-var normalizeIpp = crawler.normalizeIpp;
-var normalizePubKey = crawler.normalizePubKey;
+var argv = process.argv.slice(2);
+var rc_util = require('./rawcrawl_util.js');
+var _ = require('lodash');
 
-var rawCrawl = require('../' + process.argv.slice(2)[0]);
+if (argv.length == 1) {
+  var obj = JSON.parse(fs.readFileSync(argv[0], 'utf8'));
+  results = graphify(obj);
+  console.log(JSON.stringify(results, null, 4));
+} else {
+  console.error('eg: node misc/graphify.js misc/crawls/crawl.json');
+  process.exit(1);
+}
 
+/*
+* Returns metrics of raw crawl
+*/
 function graphify(rawCrawl) {
-
-  var out = {nodes: [], links: [], iplinks: []}
-
-  // storage
-  var nodes = []
-  var links = {}
-  var indices = {}
-
-  // Collect all unique nodes and save their indices
-  for (var ipp in rawCrawl.data) {
-    indices[ipp] = nodes.length
-    if (nodes.length == 0) {
-      nodes.push({ipp: ipp, connections: 0, entry: 1})
-    } else {
-      nodes.push({ipp: ipp, connections: 0, entry: 0})
-    }
-  }
-
-  // For each unique node
-  for(i in nodes) {
-    node = nodes[i];
-    ipp = node.ipp;
-    peers = rawCrawl.data[ipp].overlay.active;
-
-    // For each (not necessarily unique) peer
-    for (p_index in peers) {
-      peer = peers[p_index];
-
-      // Get ipp
-      try {
-        peer_ipp = normalizeIpp(peer.ip, peer.port);
-      } catch (err) {
-        peer_ipp = undefined;
-      }
-      peer_i = indices[peer_ipp];
-
-      // If it has an ipp
-      if (peer_ipp && peer_i !== undefined) {
-        peer_pk = normalizePubKey(peer.public_key);
-        peer_v = peer.version;
-        peer_t = peer.type;
-
-        // Add missing properties in node list
-        peer_node = nodes[peer_i];
-
-        //console.error(peer_ipp)
-        if (!peer_node.pk) {
-          peer_node.pk = peer_pk
-        }
-        if (!peer_node.v) {
-          peer_node.v = peer_v
-        }
-        if (!peer_node.loc) {
-          geoloc = geoip.lookup(peer_ipp.split(':')[0])
-          peer_node.loc = geoloc.country + '_' + geoloc.city
-        }
-
-        // Make link
-        var link = undefined
-        if (peer_t) {
-          // Get link
-          if (peer_t == "in") {
-            link = [i, peer_i]
-          }
-          else if (peer_t == "out") {
-            link = [peer_i, i]
-          }
-          else if (peer_t == "peer") {
-            if (peer.ip) {
-              if (peer.ip.split(":").length == 2)
-                link = [peer_i, i]
-              else
-                link = [i, peer_i]
-            }
-          }
-
-          if (link) {
-            links[link] = 1
-          }
-        }
-
-        //
-
-      } else {
-        // No ip address, can't really do much
-      }
-
+  var results = { nodes: [], 
+                  links: [] }
+
+  var pkToIndex = {};
+  var rippleds = rc_util.getRippledsC(rawCrawl.data);
+  var links = rc_util.getLinks(rawCrawl.data);
```

Comment:
I tried to abstract each util function from one another as much as possible so that they aren't dependent on one another
That's why they all take a raw crawl, even though some only need what getRippleds returns.

This makes it slower, but easier to use and somewhat safer because of abstraction.

Is it worth the tradeoff?